src/runtime/HalidePyTorchHelpers.h

```diff
@@ -0,0 +1,91 @@
+#ifndef HL_PYTORCH_WRAPPER_H
+#define HL_PYTORCH_WRAPPER_H
+
+/** \file
+ * Set of utility functions to wrap PyTorch tensors into Halide buffers,
+ * making sure the data in on the correct device (CPU/GPU).
+ */
+
+#include <exception>
+#include <iostream>
+#include <string>
+#include <sstream>
+#include <vector>
+
+#include "torch/extension.h"
+
+#include "HalideBuffer.h"
+
+#ifdef HL_PT_CUDA
+#include "HalideRuntimeCuda.h"
+#include "cuda.h"
+#include "cuda_runtime.h"
+#endif
+
+#define WEAK __attribute__((weak))
+
+#define HLPT_CHECK_CONTIGUOUS(x) AT_ASSERTM(x.is_contiguous(), #x " must be contiguous")
+#define HLPT_CHECK_CUDA(x) AT_ASSERTM(x.type().is_cuda(), #x " must be a CUDA tensor")
+#define HLPT_CHECK_DEVICE(x, dev) AT_ASSERTM(x.is_cuda() && x.get_device() == dev, #x " must be a CUDA tensor")
+
+
+namespace Halide {
+namespace PyTorch {
+
+using Halide::Runtime::Buffer;
+
+inline std::vector<int> get_dims(const at::Tensor tensor) {
+  int ndims = tensor.ndimension();
+  std::vector<int> dims(ndims, 0);
+  // PyTorch dim order is reverse of Halide
+  for (int dim = 0; dim < ndims; ++dim) {
+    dims[dim] = tensor.size(ndims-1-dim);
+  }
+  return dims;
+}
+
+
+template<class scalar_t>
+inline void check_type(at::Tensor &tensor) {
+  AT_ERROR("Scalar type ", tensor.scalar_type(), " not handled by Halide's PyTorch wrapper");
+}
+
+
+#define HL_PT_DEFINE_TYPECHECK(ctype,ttype,_3) \
+  template<> \
+  inline void check_type<ctype>(at::Tensor &tensor) { \
+    AT_ASSERTM(tensor.scalar_type() == at::ScalarType::ttype, "scalar type do not match"); \
+  }
+  AT_FORALL_SCALAR_TYPES_WITH_COMPLEX(HL_PT_DEFINE_TYPECHECK);
```

Comment:
At first I read AT_FORALL_SCALAR_TYPES_WITH_COMPLEX as being part of the HL_PT_DEFINE_TYPECHECK macro. Maybe add a blank line before and after to emphasize that it's not?

Reply:
ok

Reply:
should be more legible now